build/SharedFxInstaller.targets

```diff
@@ -9,31 +9,76 @@
+      Condition="!Exists('$(_InstallerSourceDir)$(DotnetRuntimeInstallerPrefix)-%(NonWindowsSharedFxRIDs.Identity).tar.gz')" />
   </Target>
 
-  <Target Name="GenerateTargzs" DependsOnTargets="_EnsureInstallerPrerequisites;_DownloadInstallers">
+  <Target Name="_GenerateCumulativeArchive">
     <PropertyGroup>
-      <SharedFxArchiveName>$(SharedFxInstallerName)-$(PackageVersion)-linux-x64.tar.gz</SharedFxArchiveName>
+      <WindowsArchive>false</WindowsArchive>
+      <WindowsArchive Condition="$(SharedFxPlatform.StartsWith('win'))">true</WindowsArchive>
+      <ArchiveExtension>.tar.gz</ArchiveExtension>
+      <ArchiveExtension Condition="'$(WindowsArchive)' == 'true'">.zip</ArchiveExtension>
     </PropertyGroup>
 
     <!-- Clear working directory -->
     <RemoveDir Directories="$(_WorkRoot)" />
     <MakeDir Directories="$(_WorkRoot)" />
 
     <!-- Create layout: Aspnet Runtime  -->
-    <Exec Command="tar -xzf $(SharedFxLinuxIntermediateArchiveFilePath) -C $(_WorkRoot)" />
-    <Exec Command="tar -xzf $(_InstallerSourceDir)$(DotnetRuntimeInstallerArchiveName) -C $(_WorkRoot)" />
+    <Exec
+      Command="tar -xzf $(SharedFxIntermediateArchiveFilePrefix)-$(SharedFxPlatform)$(ArchiveExtension) -C $(_WorkRoot)"
+      Condition="'$(WindowsArchive)' == 'false'"/>
+    <Exec
+      Command="tar -xzf $(_InstallerSourceDir)$(DotnetRuntimeInstallerPrefix)-$(SharedFxPlatform)$(ArchiveExtension) -C $(_WorkRoot)"
+      Condition="'$(WindowsArchive)' == 'false'"/>
+    <UnzipArchive
+      File="$(SharedFxIntermediateArchiveFilePrefix)-$(SharedFxPlatform)$(ArchiveExtension)"
+      Destination="$(_WorkRoot)"
+      Condition="'$(WindowsArchive)' == 'true'" />
+    <UnzipArchive
+      File="$(_InstallerSourceDir)$(DotnetRuntimeInstallerPrefix)-$(SharedFxPlatform)$(ArchiveExtension)"
+      Destination="$(_WorkRoot)"
+      Condition="'$(WindowsArchive)' == 'true'" />
+
+    <ItemGroup>
+      <SharedFxArchiveFiles Include="$(_WorkRoot)**\*" />
+    </ItemGroup>
+
+    <!-- Create Aspnet Runtime archive -->
+    <Exec
+      Command="tar -czf $(_InstallersOutputDir)$(SharedFxInstallerName)-$(PackageVersion)-$(SharedFxPlatform)$(ArchiveExtension) -C $(_WorkRoot) ."
+      Condition="'$(WindowsArchive)' == 'false'"/>
+    <ZipArchive
+      File="$(_InstallersOutputDir)$(SharedFxInstallerName)-$(PackageVersion)-$(SharedFxPlatform)$(ArchiveExtension)"
+      SourceFiles="@(SharedFxArchiveFiles)"
+      WorkingDirectory="$(_WorkRoot)"
+      Overwrite="true"
+      Condition="'$(WindowsArchive)' == 'true'"/>
+  </Target>
 
-    <!-- Create Aspnet Runtime tar.gz -->
-    <Exec Command="tar -czf $(_InstallersOutputDir)$(SharedFxArchiveName) -C $(_WorkRoot) ."/>
+  <Target Name="GenerateCumulativeArchives" DependsOnTargets="_EnsureInstallerPrerequisites;_DownloadInstallers">
+    <MSBuild Projects="$(MSBuildProjectFullPath)" Targets="_GenerateCumulativeArchive" Properties="SharedFxPlatform=%(NonWindowsSharedFxRIDs.Identity)" />
```

Comment:
Currently these steps only work for osx and linux archives. Unzipping dotnet runtime archives creates files with backslashes in their paths and the msbuild ziparchive task fails. I did come up with an ugly workaround, see https://github.com/aspnet/Universe/commit/62e28c6c0087d1a34003732213542df4708ebce2, but I'm sure we have a better mechanism for handling this no? Another alternative would be to create a separate build configuration that runs on windows to generate these archives.

Reply:
>  Unzipping dotnet runtime archives creates files with backslashes in their paths

Which file paths specifically? We should notify the core-setup team that they have malformed zip files.

Also, instead of https://github.com/aspnet/Universe/commit/62e28c6c0087d1a34003732213542df4708ebce2, we could add update `UnzipArchive` to normalize file slashes

Reply:
Looks like all the files contained in dotnet-runtime have this issue.
![image](https://user-images.githubusercontent.com/2030323/34685419-1850ecce-f45d-11e7-9b72-de20aa681c0a.png)

Fortunately our zips don't seem to have this problem. Who do we contact on the core-setup team for this? In the meantime should we update UnzipArchive?

Reply:
@weshaggard would know how to fix or who to ask to fix this. These zips are produced from https://github.com/dotnet/core-setup. 

In the meantime, I think we should change the `UnzipArchive` task to normalize slashes. I've seen zips with this issue before, but never has the backslash in the zip entry been intentional...it's always been a mistake.

Reply:
https://github.com/dotnet/core-setup/blob/41949b226c09874a6a15baecda6bea3184dc00f9/src/pkg/packaging/dir.proj#L127 We are just calling tar to create these, so I would expect them to be valid. Is this a recent change? Did it used to work for you guys?

Reply:
@weshaggard  the problem is with the windows .zip files. The tarballs appear ok.

![image](https://user-images.githubusercontent.com/2696087/34688850-550c4450-f468-11e7-89fe-87dd5a048075.png)

Reply:
I see what are guys using to decompress? Most windows utilities including .NET APIs will accept either slash in the archive.

Reply:
Also it isn't clear to me what problem these backslashes are causing. 

As for the issue in our zip generation I suspect we are hitting an bug in ZipArchive class which was fixed in recent versions of the framework https://docs.microsoft.com/en-us/dotnet/framework/migration-guide/mitigation-ziparchiveentry-fullname-path-separator,

Reply:
We are using System.IO.Compression.ZipArchive to decompress this zip on fairly recent .NET Core 2.1 builds.

The problem is that unzipping on Linux or macOS does not create the right directory structure. Instead of folders, you get one folder with filenames like `host\fxr\2.1.0-preview1-26108-01\hostfxr.dll`

Reply:
Ah I see you guys are extracting windows zip files on unix. I suspect that is why no one else has complained about this. Hopefully you guys can work around this but if not please file an issue on core-setup and we will see what we can do.

Reply:
Ok. https://github.com/dotnet/core-setup/issues/3577